File: 프로그래머스/3/86053. 금과 은 운반하기/금과 은 운반하기.java
import java.util.*;

class Solution {
    
    public long solution(int a, int b, int[] g, int[] s, int[] w, int[] t) {
        
        long start = 0;
        long end = (long) (10e9 * 2 * 10e5 * 2);
        long answer = 0;
        
        while(start <= end){
            long mid = (start+end)/2;
            long totalSum = 0;
            long goldSum = 0;
            long silverSum =0;
            
            for(int i=0; i<t.length; i++){
                long cnt = mid/(2*t[i]);
                if(mid%(2*t[i]) >= t[i]) cnt++;
                
                long weight = cnt*w[i];
                goldSum += Math.min(g[i], weight);
                silverSum += Math.min(s[i], weight);
                totalSum += Math.min(g[i] + s[i], weight);               
            }
            
            if(a <= goldSum && b <= silverSum && a+b <= totalSum){
                end = mid-1;
                answer = mid;
            }else{
                start = mid+1;
            }                             
        }
        
        return answer;
    }
}

File: 프로그래머스/3/86053. 금과 은 운반하기/금과 은 운반하기.js
function solution(a, b, g, s, w, t) {
    
    var start = 0;
    var end = 10e9 * 2 * 10e5 * 2;    
    var answer = 0;
    
    while(start <= end){
        var mid = Math.floor((start+end)/2);
        var sumTotal = 0;
        var sumGold = 0;
        var sumSilver = 0;
        for(let i=0; i<t.length; i++){
            var cnt = Math.round(mid/(t[i]*2))
            var weight = w[i]*cnt;
            sumTotal += Math.min(g[i] + s[i], weight);
            sumGold += Math.min(g[i], weight);
            sumSilver += Math.min(s[i], weight);
        }
        
        if(a <= sumGold && b <= sumSilver && a + b <= sumTotal){
            answer = mid;
            end = mid-1;
        }else{
            start = mid+1;
        }
    }
    
    return answer;
}